src/providers/wfs/qgsbackgroundcachedfeatureiterator.h

```diff
@@ -0,0 +1,352 @@
+ * Instances of this class may be run in a dedicated thread (QgsThreadedFeatureDownloader)
+ *
+ * This class is somewhat technical and in an ideal world would have been
+ * merged in QgsFeatureDownloader. This is not possible since QgsFeatureDownloader
+ * derives from QObject to emit signals, but the implementation of run() might
+ * also use custom signals, and should derive again from QObject, which is not
+ * possible regarding QT/MOC constraints.
+ */
+class QgsFeatureDownloaderImpl
+{
+  public:
+    QgsFeatureDownloaderImpl( QgsBackgroundCachedSharedData *shared, QgsFeatureDownloader *downloader );
+    virtual ~QgsFeatureDownloaderImpl();
+
+    /**
+     * Start the download.
+     * \param serializeFeatures whether to notify the sharedData serializer.
+     * \param maxFeatures user-defined limit of features to download. Overrides
+     *                    the one defined in the URI. Typically by the QgsWFSProvider,
+     *                    when it cannot guess the geometry type.
+     */
+    virtual void run( bool serializeFeatures, int maxFeatures ) = 0;
+
+    //! To interrupt the download. Must be thread-safe
+    void stop();
+
+    //! Emit doStop() signal
+    virtual void emitDoStop() = 0;
+
+    // To be used when new features have been received
+    void emitFeatureReceived( QVector<QgsFeatureUniqueIdPair> features );
+
+    // To be used when new features have been received
+    void emitFeatureReceived( int featureCount );
+
+    // To be used when the download is finished (successful or not)
+    void emitEndOfDownload( bool success );
+
+#if 0
+    // NOTE: implementations should copy & paste the below block
+  signals:
+    /* Used internally by the stop() method */
+    void doStop();
+
+    /* Emitted with the total accumulated number of features downloaded. */
+    void updateProgress( int totalFeatureCount );
+#endif
+
+  protected:
+    //! Progress dialog
+    QgsFeatureDownloaderProgressDialog *mProgressDialog = nullptr;
+
+    //! Whether the download should stop
+    bool mStop = false;
+
+    /**
+     * If the progress dialog should be shown immediately, or if it should be
+        let to QProgressDialog logic to decide when to show it */
+    bool mProgressDialogShowImmediately = false;
+
+    QTimer *mTimer = nullptr;
+
+    void createProgressDialog( int numberMatched );
+
+    void setStopFlag();
+    void hideProgressDialog();
+
+    void endOfRun( bool serializeFeatures,
+                   bool success, int totalDownloadedFeatureCount,
+                   bool truncatedResponse, bool interrupted,
+                   const QString &errorMessage );
+
+  private:
+    QgsBackgroundCachedSharedData *mSharedBase;
+    QgsFeatureDownloader *mDownloader;
+    QWidget *mMainWindow = nullptr;
+    QMutex mMutexCreateProgressDialog;
+};
+
+// Sorry for ugliness. Due to QgsFeatureDownloaderImpl that cannot derived from QObject
+#define CONNECT_PROGRESS_DIALOG(actual_downloader_impl_class) do { \
```

Comment:
Is there any alternative to those macros? Like templates?

Reply:
I considered that, but I'm not sure if that would work here. Templates are a pain for me :-)